src/Servers/Kestrel/Core/src/Middleware/Internal/DuplexPipeStreamAdapter.cs

```diff
@@ -74,6 +80,8 @@ public override async ValueTask DisposeAsync()
                 // the application data to the underlying stream
                 await _outputTask;
```

Comment:
Now that this is actually being called, why not just return the _outputTask directly instead of creating a new AsyncStateMachine?

Reply:
It could have been that way from the get go actually

Reply:
Ah it would need to look like this:

`return new ValueTask(_outputTask ?? Task.CompletedTask);`

Reply:
Probably slightly better to do `return _outputTask != null ? new ValueTask(_outputTask) : default;` even though it looks uglier.

Reply:
`_outputTask?.IsCompletedSuccessfully == true ? new ValueTask(_outputTask) : default`

Reply:
You have the logic backwards. You want to return the Task-derived ValueTask when it hasn't already completed successfully, not when it has.

```C#
return _outputTask?.IsCompletedSuccessfully == false ? new ValueTask(_outputTask) : default;
```

The above actually works believe it or not. It only returns the non-default ValueTask when the _outputTask is both non-null and not successfully completed. 

I *almost* recommend it out of my love for code golf, but I didn't want to require ever reading the code to have a deep understanding of Nullable.Equals() to understand it. The fact that even you got it backwards reaffirms my choice. 😆

Reply:
I don't know what you mean (It's not like I can edit the comment)